### PR TITLE
Small object buffer fix

### PR DESCRIPTION
**Pull request by jackyalbo**
### Explain the changes
1. Adding multi_buffer_pool - a pool consisting of 4 different buffer pools with different sizes as was proposed in https://github.com/noobaa/noobaa-core/issues/7449

### Issues: Fixed #xxx / Gap #xxx
1. 

### Testing Instructions:
1. I ran this in order to see if we get improvement 

```warp get --host=127.0.0.1:6001 --access-key=jVK45OKEXls2GeEpz67G --secret-key=VRXfwvxwKfpnRb4koqsCHDW1YnQHJNPZoeqxcOt3 --disable-multipart  --noclear --duration 1m --concurrent=20 --objects=1000 --obj.size 80K --list-existing```

MASTER results
```
----------------------------------------
Operation: GET. Concurrency: 20
* Average: 99.35 MiB/s, 1302.17 obj/s

Throughput, split into 59 x 1s:
 * Fastest: 106.9MiB/s, 1400.98 obj/s
 * 50% Median: 104.7MiB/s, 1372.23 obj/s
 * Slowest: 39.0MiB/s, 510.91 obj/s
```
PR results: (20% better)
```
----------------------------------------
Operation: GET. Concurrency: 20
* Average: 120.07 MiB/s, 1573.84 obj/s

Throughput, split into 59 x 1s:
 * Fastest: 129.4MiB/s, 1695.97 obj/s
 * 50% Median: 126.3MiB/s, 1655.10 obj/s
 * Slowest: 40.5MiB/s, 530.49 obj/s
```

- [ ] Doc added/updated
- [x] Tests added

warp suggestions:
```
WARP_GET="warp get --host=127.0.0.1:6001 --access-key=jVK45OKEXls2GeEpz67G --secret-key=VRXfwvxwKfpnRb4koqsCHDW1YnQHJNPZoeqxcOt3 --disable-multipart  --noclear --duration 5m"

# step 1 - create dataset and bench get right after
$WARP_GET --concurrent=20 --objects=300 --obj.size 8M

# step 2 - bench get and read existing dataset
$WARP_GET --concurrent=20 --objects=300 --obj.size 8M --list-existing


$WARP_GET --concurrent=100 --objects=1000 --obj.size 1M
$WARP_GET --concurrent=100 --objects=1000 --obj.size 1M --list-existing


$WARP_GET --concurrent=100 --objects=1000 --obj.size 4K
$WARP_GET --concurrent=100 --objects=1000 --obj.size 4K --list-existing

```

### CODE REVIEW

**Review by guymguym — CHANGES_REQUESTED** [review]

File: src/server/bg_services/semaphore_monitor.js [hunk 1]
```diff
@@ -72,8 +72,8 @@ class SemaphoreMonitor {
         }
     }
 
-    sample_nsfs_semaphore() {
-        const buffers_pool_sem = buffers_pool.sem;
+    sample_nsfs_l_semaphore() {
```
Comment (guymguym) [inline src/server/bg_services/semaphore_monitor.js:75]: how hard it is to have all semaphores in the stats?
I see in line 92 we call:
```
endpoint_stats_collector.update_semaphore_state(semaphore_report, "nsfs", this.report_sample_sizes);
```
perhaps just cann this once per buffer_pool with different names?

**Review by guymguym — CHANGES_REQUESTED** [review]

File: src/server/bg_services/semaphore_monitor.js (hunk 1 above)
Comment (jackyalbo) [reply src/server/bg_services/semaphore_monitor.js]: done, 4 new semaphores are now supported.

File: src/util/buffer_utils.js [hunk 2]
```diff
@@ -227,6 +229,52 @@ class BuffersPool {
...
+            size,
+            buffers_pool: new BuffersPool({
+                buf_size: size,
+                sem: new Semaphore(sem_size,
+                {
+                    timeout: sem_timeout || config.IO_STREAM_SEMAPHORE_TIMEOUT,
+                    timeout_error_code: sem_timeout_error_code || 'IO_STREAM_ITEM_TIMEOUT',
+                    warning_timeout: sem_warning_timeout || config.NSFS_SEM_WARNING_TIMEOUT,
+                }),
+                warning_timeout: warning_timeout || config.NSFS_BUF_POOL_WARNING_TIMEOUT,
```
Comment (guymguym) [inline src/util/buffer_utils.js:257]: remove specific nsfs defaults from util code

File: src/util/buffer_utils.js [hunk 3]
```diff
@@ -227,6 +229,52 @@ class BuffersPool {
...
+    get_buffers_pool(size) {
+        let buffer_pool;
+        for (let i = 0; i < this.pools.length; ++i) {
+            buffer_pool = this.pools[i].buffers_pool;
+            if (size <= this.pools[i].size) {
+                break;
+            }
+        }
+        dbg.log1('MultiSizeBuffersPool.get_pool: sem value', buffer_pool.sem._value, 'waiting_value', buffer_pool.sem._waiting_value, 'buffers length', buffer_pool.buffers.length);
+        return buffer_pool;
```
Comment (guymguym) [inline src/util/buffer_utils.js:275]:
```suggestion
        for (const bp of this.pools) {
            if (size <= bp.buf_size) {
                dbg.log1('MultiSizeBuffersPool.get_pool: sem value', bp.sem._value, 'waiting_value', buffer_pool.sem._waiting_value, 'buffers length', buffer_pool.buffers.length);
                return bp;
            }
        }
        return this.pools[this.pools.length - 1]; // largest
```

File: src/sdk/namespace_fs.js [hunk 4]
```diff
@@ -1527,7 +1534,8 @@ class NamespaceFS {
             const pre_known_size = (params.size >= 0);
             if (!pre_known_size) {
                 // 2.1
-                upload_res = await buffers_pool_sem.surround_count(config.NSFS_BUF_SIZE, async () => this._upload_stream(md_upload_params));
+                const bp = multi_buffer_pool.get_buffers_pool(config.NSFS_BUF_SIZE_L);
+                upload_res = await bp.sem.surround_count(config.NSFS_BUF_SIZE_L, async () => this._upload_stream(md_upload_params));
```
Comment (guymguym) [inline src/sdk/namespace_fs.js:1538]: we want to avoid reading the config dynamically and instead use the largest pool.
get_buffers_pool(undefined) can be used for that.

File: src/util/chunk_fs.js [hunk 5]
```diff
@@ -55,13 +55,13 @@ class ChunkFS extends stream.Transform {
             });
             this.count = 0;
             while (chunk && chunk.length) {
-                const available_size = config.NSFS_BUF_SIZE - this.q_size;
+                const available_size = config.NSFS_BUF_SIZE_L - this.q_size;
```
Comment (guymguym) [inline src/util/chunk_fs.js:58]: we want to avoid reading the config dynamically and instead use the largest pool.
```
const { buf_size } = get_buffers_pool(undefined);

...
const available_size = buf_size - this.q_size;
```

File: src/sdk/namespace_fs.js [hunk 6]
```diff
@@ -1111,10 +1118,10 @@ class NamespaceFS {
             upload_params = await this._start_upload(fs_context, object_sdk, file_path, params, open_mode);
 
             if (!params.copy_source || upload_params.copy_res === copy_status_enum.FALLBACK) {
-            // TODO: Take up only as much as we need (requires fine-tune of the semaphore inside the _upload_stream)
-            // Currently we are taking config.NSFS_BUF_SIZE for any sized upload (1KB upload will take a full buffer from semaphore)
-                const upload_res = await buffers_pool_sem.surround_count(
-                    config.NSFS_BUF_SIZE, async () => this._upload_stream(upload_params));
+            // We are taking the buffer size closest to the sized upload
+                const bp = multi_buffer_pool.get_buffers_pool(params.size >= 0 ? params.size : config.NSFS_BUF_SIZE_L);
```
Comment (guymguym) [inline src/sdk/namespace_fs.js:1122]: we want to avoid reading the config dynamically and instead use the largest pool.
use get_buffers_pool(params.size) and handled it inside the call to return the largest pool in that case.

File: src/test/unit_tests/test_buffer_pool.js [hunk 7]
```diff
@@ -44,4 +44,53 @@ mocha.describe('Test buffers pool', function() {
         assert(buffers_pool.sem.value === SEM_LIMIT, 'Smepahore did not deallocate after buffers release');
     });
 
+    mocha.it('Work parallel with buffers with different sizes and different semaphores', async function() {
```
Comment (guymguym) [inline src/test/unit_tests/test_buffer_pool.js:47]: add test case for undefined and negative sizes passed to get_buffer_pool that return the largest, but 0 should return the smallest.

**Review by guymguym — APPROVED** [review]

File: config.js [hunk 8]
```diff
@@ -658,8 +659,29 @@ assert(config.NAMESPACE_CACHING.DEFAULT_BLOCK_SIZE > config.INLINE_MAX_SIZE);
...
-config.NSFS_BUF_SIZE = 8 * 1024 * 1024;
-config.NSFS_BUF_POOL_MEM_LIMIT = config.BUFFERS_MEM_LIMIT;
+config.NSFS_BUF_SIZE_XS = 4 * 1024;
+config.NSFS_BUF_SIZE_S = 64 * 1024;
+config.NSFS_BUF_SIZE_M = 1 * 1024 * 1024;
+config.NSFS_BUF_SIZE_L = 8 * 1024 * 1024;
+
+config.NSFS_WANTED_BUFFERS_NUMBER = 512;
+config.NSFS_MAX_MEM_SIZE_XS = 8 * 1024 * 1024;
+config.NSFS_MAX_MEM_SIZE_S = 32 * 1024 * 1024;
```
Comment (guymguym) [inline config.js:669]: add explanation on these because it is fresh in your memory now!

File: src/sdk/namespace_fs.js [hunk 9]
```diff
@@ -447,7 +454,7 @@ class NamespaceFS {
         force_md5_etag,
     }) {
         dbg.log1('NamespaceFS: buffers_pool length',
-            buffers_pool.buffers.length, buffers_pool.sem);
+            multi_buffer_pool.buffer_pools);
```
Comment (guymguym) [inline src/sdk/namespace_fs.js:457]: This prints undefined. We want to print the info of all the pools, so we best add a custom inspect method to buffer pool - see https://nodejs.org/dist/latest-v20.x/docs/api/util.html#custom-inspection-functions-on-objects
